### ISSUE TITLE
Move mailing list to Mailman 3 at {lists}.operate-first.cloud

### ISSUE THREAD

**Issue by quaid** [operate-first/community#53]
Recommendation:  move the public mailing list to a neutral Operate First community branded URL (operate-first.cloud), migrating to Mailman 3 in the process with it's modern web forum UI.

Alternative:  move the public mailing list to a neutral openinfralabs.org URL. May not be able to do Mailman 3 migration as they are running MM 2, which is the same major version as listman.redhat.com.

Background
=========
Communication in an open source project benefits from an openly archived, easy to use threaded discussion platform. The classic solution is a threaded mailing list with public archives and their all-so-valuable human readable and fixed URLs for messages and discussion threads. The real-time chat interfaces with archiving have not been able to replace this functionality. Arguably the threaded discussion experience is either sufficient or superior, but the access to public archives and easy linking is not.

As it happens, we have a project mailing list, but it is underused. I reckon that not having a web interface and being on a redhat.com domain are two things that might affect adoption.

Aside from the value in conducting and archiving threaded discussions of decision making, an openly archived mailing list is a common place to make an announcement such as about a release, a feature being deprecated, part of the project restructuring itself, and any other piece of information you want the project and the general public to be able to easily read.

**Comment by quaid** [comment]
This is an active project to maintain Mailman 3 container images:

https://github.com/maxking/docker-mailman

Some how-to instructions:  https://asynchronous.in/docker-mailman/

**Comment by quaid** [comment]
@durandom Who should Misc (OSCI sysadmin) get in contact with on the ops side? It's probably going to be things around DNS, Let's Encrypt certs, and the like.

@mscherer What do you need for the phase 1 mailing list setup?

**Comment by durandom** [comment]
I guess we re-use the current installation of mailman at https://osci.io/mission_and_services/ ?
If that's the case, you'd need DNS entries?

@goern knows how to create these. Maybe @operate-first/admins also know more.
In anyway, we should open an issue with concrete asks at https://github.com/operate-first/support

**Comment by quaid** [comment]
> I guess we re-use the current installation of mailman at https://osci.io/mission_and_services/ ?
> If that's the case, you'd need DNS entries?

Yes to DNS. But aiui we'll have our own uniquely IP'd VM to host the subdomain. I'm pretty sure it's all run out of https://gitlab.com/osci and so there may be additional automation we can do from Operate First GitOps.

> @goern knows how to create these. Maybe @operate-first/admins also know more.
> In anyway, we should open an issue with concrete asks at https://github.com/operate-first/support

@mscherer Can you create the issue as you know all the details?

**Comment by mscherer** [comment]
To be clear, the list server would be hosted in a RH Datacenter (in a separate zone, no contact with internal lan) in Raleigh (Morrisville to be precise). We provides the VM, the update of the hypervisors, etc.

But before deploying, I need to know:
- the server name ? I guess lists.operate-first.cloud, but sometime, people want something else (and this can't be changed easily, we have to reinstall basically)
 
- what kind of authentication you want (eg, github, others), and associated secrets. It can be added later.
- who will moderate the lists (can also be added later)
- if you need a specific theme/colorscheme/etc. Can be added later, but we do not do the design job (unless you want people scream when they see my "css art"). 
- the lists of lists to create, but again, can be done later

So In short, what is the DNS name that should be used ? I will open tickets with that, and deal with the rest.
And we need also contacts (so folks to warn when we do maintainance, etc). 2 names should be enough.

**Comment by durandom** [comment]
* `lists.operate-first.cloud` is good. if that's a canoncial name for lists.
* github is good. that'll be the primary source for the SSO too

cc @tumido @HumairAK

**Comment by quaid** [comment]
Let's add Google for oauth, to increase the ease with which people join. (They can always make a stand-alone account, yes, so this is more about the appearance of being easy.)

Moderators, themes, and specific lists we can add later. I should be on the contact list for maintenance, etc. Do we have an admin email alias like @operate-first/admins ?

**Comment by mscherer** [comment]
We need someone with DNS access to set lists.operate-first.cloud to 8.43.85.122 
I am installing the server at the moment, but the rest is blocked on the DNS (because we use let's encrypt and ansible).

**Comment by durandom** [comment]
I created an issue at https://github.com/operate-first/support/issues/348

**Comment by mscherer** [comment]
For email alias, if the team do not want to bother maintaining a smtp relay, that's also something my team can do.

**Comment by quaid** [comment]
> For email alias, if the team do not want to bother maintaining a smtp relay, that's also something my team can do.

Seems like a good idea to have the OSCI team provide smtp relay for now.

My architect guidance is, let's start with OSCI providing a fully contained solution, i.e., have your team run the SMTP relay and anything else that might come up. Even if we duplicate a service to start, we can resolve that once the service is running. Then the next phase is discussing how and when to move what to where. :)

**Comment by mscherer** [comment]
So https://lists.operate-first.cloud/ is up. I need @quaid to create a account so I can promote the account to admin, and then we have:
- get a github secret (so the system will say "operate-first org ask you access to"...)
- get a google secret

I can help (read guide a bit) for the github one, I have done several time, but for the google one, I might be a bit less helpful. We have a osci-rh gmail account just for that case, but really, it is mostly to decide how much people will trust some random name asking for credential access.

**Comment by tumido** [comment]
Can we integrate this with Operate First SSO instead of custom google/github auths? You can create a Keycloak client in here:
https://github.com/operate-first/apps/tree/master/keycloak/overlays/moc/infra/clients

We're in process of migrating our existing clusters to this Keycloak so we can have a true Single SO.

**Comment by tumido** [comment]
Btw, why don't we deploy the server within Operate First as a containerized application on OCP?

**Comment by quaid** [comment]
> Can we integrate this with Operate First SSO instead of custom google/github auths? You can create a Keycloak client in here:
> https://github.com/operate-first/apps/tree/master/keycloak/overlays/moc/infra/clients
> 
> We're in process of migrating our existing clusters to this Keycloak so we can have a true Single SO.

I think that makes sense from a high level for sure, some questions about how that effects this rollout:

* How long might this delay rollout of the list server?
* How can we improve the UX for the SSO tool?

For the latter I mean, I presume you are referring to the login window for OCP at MOC? The UX for that is confusing for people who don't already know what is going on. Whereas the GitHub and Google oAuth experiences are something people are familiar with, the OCP login is not a familiar experience. It is Red Hat branded, says nothing about Operate First, has two buttons that only make sense to people who already know the system architecture (`kube:admin` and `moc-sso`), and effectively makes the user feel like they left the website they started from.

![OCP_login-UX](https://user-images.githubusercontent.com/112760/129588904-7621fcd0-5c09-4e46-a9ef-03ac0a0ab81d.png)

So if our intention is to roll that out system wide, how can I get involved in what the community user experience is going to be like?                                                                                                                                                                                                                                                                                                            

> Btw, why don't we deploy the server within Operate First as a containerized application on OCP?

That indeed is the plan!

In terms of, "Why not wait until we can do this the correct way?", the reason is almost half of my community initiatives are blocked on not having the mailing list:

https://github.com/operate-first/community/blob/main/program-management-tracker.md

Once we're finished with this phase, we can open a discussion on the new mailing list to discuss doing that. :)

**Comment by quaid** [comment]
> So https://lists.operate-first.cloud/ is up. I need @quaid to create a account so I can promote the account to admin, and then we have:

I created one for 'quaid' and 'quaid@iquaid.org', I haven't received my activation email yet.

>     * get a github secret (so the system will say "operate-first org ask you access to"...)
> 
>     * get a google secret
> 
> 
> I can help (read guide a bit) for the github one, I have done several time, but for the google one, I might be a bit less helpful. We have a osci-rh gmail account just for that case, but really, it is mostly to decide how much people will trust some random name asking for credential access.

Let's look to see what effort the SSO approach would be.

I reckon we don't want to remove an auth path once we've created it, right? Or else I'd say let's go ahead with one and change later.

We also don't want to ask people to create a new account so they can SSO into our system, so we need to make sure that the SSO offers multiple auth providers, including GitHub.

We might want to just offer multiple paths for people so they can choose and we don't have to manage their accounts. So, that means we add auth sources when we decided they are trustworthy and it's worth it to us.

We could start with GitHub, and maybe the Op1st SSO can provide us the Google auth pathway.

**Comment by mscherer** [comment]
@tumido The problem is that mailman 3 is not just "1 server". Mailman 3 as we deploy requires postgres, a python server (mailman itself), a django web interface (hyperkitty in django, so httpd), memcached, its own queing system. It also requires postfix, and we deploy it with postgrey, and spamassasin. 

The whole system also communicate with unix sockets and/or non HTTP TCP protocols, and at least mailman and postfix are starting and stopping their own processes (which is IMHO duplicating the same job as k8s/openshift, and not exactly how we would do in a more cloud native world).

This is doable, but not trivial, and no one did the work (even if we had a intern who worked on the postfix part). 

My team plan was to look at doing that using kubevirt and slowly see how we can move part of the system outside. 

Also, when it come to SMTP, having a fixed and good IP can be quite important. Having mailman on openshift (in a clean way) is a interesting endeavor, but I suspect it take at least a few weeks to get it running.

**Comment by tumido** [comment]
Sure.. :) Let's start with the custom deployment and move towards Operate first driven one :slightly_smiling_face: 

When it comes to SSO, I'm referring to the flow that happens **after** you click on "operate-first" button on this screen. Mind we don't have Operate First SSO enabled in MOC yet. So far only on Rick in EMEA.

https://console-openshift-console.apps.rick.emea.operate-first.cloud/
![image](https://user-images.githubusercontent.com/7453394/129716835-16fb1499-5751-4e2b-a2e3-ccd300fbbd6f.png)
 
On the screen you can see UI that comes from OpenShift. our SSO has auto forwards everywhere so user doesn't even see it and is redirected directly to the GitHub auth:
![image](https://user-images.githubusercontent.com/7453394/129717024-4c77833c-32ae-4d24-bbbb-ee5f82e73ef2.png)

That's it.

**Comment by tumido** [comment]
@quaid ^

I'd really prefer if you can integrate with Operate First SSO instead of a custom solution. It's really easy to provision a client for you and then you just OIDC your app with client id and secret you set in the `KeycloakClient` resource, your own redirect url into the app and our SSO host. That's it.

Also our users can then see the mailing list app in the applications tab on their profile etc.
https://keycloak-keycloak.apps.moc-infra.massopen.cloud/auth/realms/operate-first/account/
![image](https://user-images.githubusercontent.com/7453394/129717502-47db0230-c19c-4b2c-85b0-b84f8bd3e084.png)

**Comment by mscherer** [comment]
We can get several login solutions, that's not a problem (eg github and internal SSO and login/password and Google). 

And I think having the most popular options is usually good. If you ask to external people to create a account somewhere before they can ask questions or join, you will lose a portion of future productive contributors. Just think each time you had to subscribe to yet another bug tracker to report a problem, and how annoyed you were. 

I am not advocating to use all the backend possible, but here is the list of what can be done:
https://python-social-auth.readthedocs.io/en/latest/backends/index.html#supported-backends

So if you want SAML auth, sure, this can be added along the others.

If tomorrow, you discover that the community is mostly Untappd users, then we ca add the backend. 
If the day after, China is where we want to get people from (and they do not have github accounts, nor google ones), I guess there is options too.

**Comment by quaid** [comment]
> I'd really prefer if you can integrate with Operate First SSO instead of a custom solution. It's really easy to provision a client for you and then you just OIDC your app with client id and secret you set in the `KeycloakClient` resource, your own redirect url into the app and our SSO host. That's it.

Thanks, I followed that login path as described. I'm glad to see it going back into use the more familiar GitHub account experience.

@mscherer This is a path to getting GitHub enabled within existing systems. Because Mailman 3 lets you create a stand-alone account in their system instead, we are not requiring a GitHub account to participate. We'll want to advertise that on the lists description page.

@tumido Is there a way to change the styling and/or branding on the login window? The links text is a bit obscure, can we add any context around it?

The plan is to migrate over all the existing subscribers, but I don't want to do that if they can't self-service their own account, or if doing so means they have to create stand-alone accounts.

Can we put up the Operate First community SSO today (18 Aug)?

**Comment by durandom** [comment]
> So if you want SAML auth, sure, this can be added along the others.

Does that mean mailman can auth against keycloak? If technically possible, I'd like to 

1. subscribe without auth, unsubscribe without auth (good old mailinglist management)
2. auth against Operate First SSO (i.e. Keycloak), if you like access to web based management

This is the lowest entry for getting access to the ML and the least maintenance if it comes to auth management.

**Comment by quaid** [comment]
> > So if you want SAML auth, sure, this can be added along the others.
> 
> Does that mean mailman can auth against keycloak? If technically possible, I'd like to
> 
>     1. subscribe without auth, unsubscribe without auth (good old mailinglist management)
> 
>     2. auth against Operate First SSO (i.e. Keycloak), if you like access to web based management
> 
> 
> This is the lowest entry for getting access to the ML and the least maintenance if it comes to auth management.

I concur, but I don't know if Mailman can auth against keycloak.

@mscherer @tumido Can you try to set this up before the end of the week? So we can at least know if it's going to be easy or not?

**Comment by durandom** [comment]
Can't we start with `1. subscribe without auth, unsubscribe without auth (good old mailinglist management)` ?
I think that's already implemented. Then that wont block the new ML launch.

**Comment by quaid** [comment]
To be clear, I'm asking if we can start on making the connection and see how complex it appears to be.

If it's too complex or seems like it may take another four weeks, then I wouldn't want to wait and would proceed with just the 1. auth method for now.

> Can't we start with `1. subscribe without auth, unsubscribe without auth (good old mailinglist management)` ?
> I think that's already implemented. Then that wont block the new ML launch.

My concern is this: 

We are migrating subscribers from an old list to a new one, but we aren't sure they all are aware of what is going on if they haven't read the threads on the topic. Force subscribing people to a mailing list is frowned upon unless they are aware of it; the alternative of sending out subscription invites is a sure way to make sure that you lose half of your subscribers in a migration. (And half of those lost didn't want to be lost, but didn't notice or understand the "do you want to subscribe" email.)

In such imperfect migration situations, the first email I send to the new list is a reminder of what just happened and why, and tells people how to unsubscribe if they no longer wish to be on the list. (I'll also offer to simply remove or change subscriptions for anyone who asks, as a courtesy.)

But in order to subscribe, one has to login. That means counterintuitively creating an account for a system that already has a record for you as a user.

So using the native good old mailinglist management is a bit less smooth than "Login via Operate First SSO with your GitHub account."

Now, we can certainly go ahead and use just the 1. auth method and a bit of extra how-to material. But if the keycloak is straightforward to configure, then I'd rather wait a day or two for that. If it won't be a day or two, then let's proceed without it for now and add it when it's ready.

**Comment by mscherer** [comment]
Mailman 3 support SAML, OAuth, OpenID. Keycloak implement Samal, Oauth and OpenID Connect, so that's 2 out of 3. 

Assuming we chose SAML, the required information are listed in the doc:
https://python-social-auth.readthedocs.io/en/latest/backends/saml.html#required-dependency

So I need to generate a cert, etc, but I also need the information to fill SOCIAL_AUTH_SAML_ENABLED_IDPS

**Comment by mscherer** [comment]
Ok so I think SAML is "niet". The host is EL 7 (since mailman stack is not yet packaged for EL 8), and there is no python saml packages, as far as I see. The documentation is also a lot less straightforward than github and others, and while there is nothing that can't be packaged nor fixed, it would take some time.

I will see how Fedora do, since they use freeipa as their auth backend now, not sure if that's using openid connect,  oauth or saml.

**Comment by quaid** [comment]
Thanks for looking into it @mscherer, that's just what I needed to know.

I'll go ahead with the mass subscription and an appropriate "welcome" email.

That resolves this issue, I'll open a new one just for exploring auth options, since that is now a discrete task no longer associated with this move and migration.

I'm pretty stoked with how this collaboration went, I'm looking forward to the next one. Thanks y'all!
